### PR TITLE
feat(cli): add `--retry [retries]` to `wing test`

### LINKED ISSUE #3796: adding a `retry` option to the cli test command
### Feature Spec

Adding a `retry` option to the cli test command to re-run the failed tests for a specific amount of times.

### Use Cases

Mostly for automated tests that tend to fail over environmental factors rather than the code that is tested. It can save plenty of time, since now we're repeating the command, it runs again all the tests- including the successful ones.

### Implementation Notes

_No response_

### Component

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by garysassano**
Closes #3796

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: tools/hangar/__snapshots__/test_corpus/valid/print.test.w_test_sim.md [hunk 1]
```diff
@@ -8,6 +8,8 @@ pass ┌ print.test.wsim » root/env0/test:log1
      │ inflight log 1.1
      └ inflight log 1.2
 pass ┌ print.test.wsim » root/env1/test:log2
+     │ inflight log 1.1
+     │ inflight log 1.2
```
Comment (MarkMcCulloh) [inline tools/hangar/__snapshots__/test_corpus/valid/print.test.w_test_sim.md:12]: Looks like the logs are getting messed up now that the tests are running in parallel.

Maybe keep them running serially for this PR and work on parallelizing on a different one.

File: apps/wing/src/cli.ts [hunk 2]
```diff
@@ -191,6 +191,11 @@ async function main() {
       "-o, --output-file <outputFile>",
       "File name to write test results to (file extension is required, supports only .json at the moment)"
     )
+    .addOption(
+      new Option("-R, --retry [retries]", "Number of times to retry failed tests")
+        .preset(3)
```
Comment (MarkMcCulloh) [inline apps/wing/src/cli.ts:196]: Not sure if we want this as a default. During development iteration, test failures will be normal and retying them will just lead to noise.

Comment (garysassano) [reply apps/wing/src/cli.ts:196]: What do you mean "as a default"? That's only if you use the `--retry` option in the command.

Comment (garysassano) [reply apps/wing/src/cli.ts:196]: I could probably add an extra test for when the `--retry` parameter isn't passed, to check that it actually doesn't retry failed tests if you want.

Comment (garysassano) [reply apps/wing/src/cli.ts:196]: If you want, I could probably add an extra test for when the --retry parameter isn't passed, to check that it actually doesn't retry failed tests.

Comment (garysassano) [reply apps/wing/src/cli.ts:196]: If you want, I could probably add an extra test for when the `--retry` parameter isn't passed, to check that it actually doesn't retry failed tests.

Comment (garysassano) [reply apps/wing/src/cli.ts:196]: To sum it up:
- `wing test [...endpoint]` = 0 retries
- `wing test [...endpoint] --retry` = 3 retries
- `wing test [...endpoint] --retry x` = x retries

Comment (MarkMcCulloh) [reply apps/wing/src/cli.ts:196]: Gotcha, I misread it, yeah then that makes sense

File: apps/wing/src/commands/test/test.ts [hunk 3]
```diff
@@ -29,11 +29,23 @@ const generateTestName = (path: string) => path.split(sep).slice(-2).join("/");
...
+   */
   outputFile?: string;
-  /** String representing a RegEx used for test filtering. */
+  /**
+   * String representing a regex pattern used to selectively filter which tests to run.
+   */
   testFilter?: string;
+  /**
+   * How many times failed tests should be retried.
+   * @default 3
```
Comment (MarkMcCulloh) [inline apps/wing/src/commands/test/test.ts:46]: In the context of calling test(), the default is actually 0
```suggestion
   * @default 0
```

**Review by MarkMcCulloh — APPROVED** [review]

File: apps/wing/src/cli.ts (hunk 2 above)
Comment (garysassano) [reply apps/wing/src/cli.ts:196]: For the sake of completeness, I've added a test for when the option is missing.

File: apps/wing/src/commands/test/test.ts (hunk 3 above)
Comment (garysassano) [reply apps/wing/src/commands/test/test.ts]: You are right, if you call `test()` directly the default value would be indeed `0`. It's Commander that sets the value to `3` when calling the method through it.

Comment (garysassano) [reply apps/wing/src/commands/test/test.ts]: You are right, if you call `test()` directly the default value would be indeed `0`. It's Commander that sets the value to `3` or `x` when calling the method through it.

Comment (garysassano) [reply apps/wing/src/commands/test/test.ts]: Actually, there's no default value. If you call `test()` directly, the `retry` variable would be undefined. It's Commander that sets the value but that doesn't exist without it.

Comment (garysassano) [reply apps/wing/src/commands/test/test.ts]: It's the reason why I manually set the value to `0` here: https://github.com/winglang/wing/pull/4666/files#diff-cc35dcb1d186df3226515f587ca678d51e6c8e1af1ef38684e312d30bbcab5daR286

**Pull request merged**